activesupport/lib/active_support/core_ext/hash/deep_compact.rb

```diff
@@ -0,0 +1,49 @@
+class Hash
+  def deep_compact(blank: false)
+    dup.deep_compact!(blank: blank)
+  end
+
+  def deep_compact!(blank: false)
+    each_key do |k|
+      v = self[k]
+
+      case v
+      when Hash
+        v.deep_compact!(blank: blank)
+        delete(k) if blank ? v.blank? : v.nil?
+      when Array
+        v.map! do |elem|
+          if elem.is_a?(Hash)
+            elem.deep_compact!(blank: blank)
+            elem
+          elsif elem.is_a?(Array)
+            deep_compact_array(elem, blank)
+          else
+            elem
+          end
+        end
+        v.reject! { |elem| blank ? elem.blank? : elem.nil? }
+        delete(k) if blank ? v.blank? : v.nil?
+      else
+        delete(k) if blank ? v.blank? : v.nil?
+      end
+    end
```

Comment:
You can probably remove the duplication here since it applies to all cases:

```suggestion
      case v
      when Hash
        v.deep_compact!(blank: blank)
        
      when Array
        v.map! do |elem|
          if elem.is_a?(Hash)
            elem.deep_compact!(blank: blank)
            elem
          elsif elem.is_a?(Array)
            deep_compact_array(elem, blank)
          else
            elem
          end
        end
        v.reject! { |elem| blank ? elem.blank? : elem.nil? }
      end
    end
    delete(k) if blank ? v.blank? : v.nil?
```

Reply:
Thanks, I've done some refactoring. Feel free to check it out again.